Add Telephone migration functionality

bkjohnson opened the pull request:
## Description

Closes https://github.com/department-of-veterans-affairs/va.gov-team/issues/34915

The current component migration script completely removes the React component import:

https://github.com/department-of-veterans-affairs/vets-website/blob/96d509701c079064f5e3a358d519567378100469/script/component-migration/index.js#L116

For migrating `<Telephone>` we don't want to do that since we still want to leave the `CONTACTS` import. Here's an example:

https://github.com/department-of-veterans-affairs/vets-website/blob/39312af1c9f48ca14c453d79417452eb99fd88f4/src/platform/static-data/SubmitSignInForm.jsx#L3-L6

Rather than further tweaking the regex-based migration script to handle this kind of case, we decided to explore a custom ESLint rule. This new rule also offers a suggestion that a developer can choose to apply right from their editor.

Configuring the rule to be autofixable causes issues with the current husky config which autofixes any ESLint issues for changed files right when a developer is about to commit. We don't want these migrations to be applied that way, so the suggestion allows us to warn and give an easy way to migrate.

## Original issue(s)
https://github.com/department-of-veterans-affairs/va.gov-team/issues/34915


## Testing done
- Unit testing
    - `yarn test:unit script/eslint-plugin-va/`
- Local runs of ESLint on select directories to show warnings
- Testing the suggestion integration with VS code


## Screenshots

Using VS Code's ESLint plugin to manually apply the suggestions:

![Peek 2022-01-26 15-40](https://user-images.githubusercontent.com/2008881/151265817-5a865f06-f9b5-4f74-96b5-6aebb189fba6.gif)

### Changed file

- `contact` remains, possibly stripped of hyphens
- `extension` remains
- `pattern` is removed
- `international` is added if `pattern` was `PATTERNS.OUTSIDE_US`

![image](https://user-images.githubusercontent.com/2008881/150889668-0ac32b17-8c39-4ace-a653-bb362456f213.png)

## Acceptance criteria
- [ ]

## Definition of done
- [ ] Events are logged appropriately
- [ ] Documentation has been updated, if applicable
- [ ] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [ ] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

bkjohnson committed:
Link to local eslint plugin

This makes testing easier since I don't have to yarn install after each
change to see the results

bkjohnson committed:
Add initial fixable rule configuration

Will improve to make this more extendable

bkjohnson committed:
Move component work into separate function

bkjohnson committed:
Add international prop when PATTERN suggests it

bkjohnson committed:
Migrate `notClickable` to `not-clickable`

bkjohnson committed:
Add helper function

bkjohnson committed:
Add sample file

bkjohnson committed:
Use optional chaining for international

bkjohnson committed:
Fix a typo

bkjohnson committed:
Include an actual message for the lint result

bkjohnson commented on script/eslint-plugin-va/rules/prefer-web-component-library.js:
```diff
@@ -0,0 +1,60 @@
+  context.report({
+    node,
+    message: MESSAGE,
+    data: {
+      reactComponent: componentName.name,
+      webComponent: 'va-telephone',
+    },
```
This is just to [fill in the message placeholders](https://eslint.org/docs/developer-guide/working-with-rules#using-message-placeholders)

bkjohnson commented on script/eslint-plugin-va/rules/prefer-web-component-library.js:
```diff
@@ -0,0 +1,60 @@
+      reactComponent: componentName.name,
+      webComponent: 'va-telephone',
+    },
+    fix: fixer => {
+      // Replace the node name
+      // and remove the `pattern` prop if it's there
+      return [
+        fixer.replaceText(componentName, 'va-telephone'),
+        international
+          ? fixer.insertTextBefore(patternNode, 'international')
+          : null,
+        patternNode ? fixer.remove(patternNode) : null,
+        notClickableNode
+          ? fixer.replaceText(notClickableNode, 'not-clickable')
+          : null,
+      ].filter(i => !!i);
```
This is the equivalent to what the transformer function would do if we implemented this as part of the original migration script.

bkjohnson commented on script/eslint-plugin-va/rules/prefer-web-component-library.js:
```diff
@@ -0,0 +1,60 @@
+
+  create(context) {
+    return {
+      JSXElement(node) {
+        const componentName = node.openingElement.name;
+        if (componentName.name === 'Telephone') {
+          telephoneTransformer(context, node);
+        }
+      },
```
If we go with this linting rule, this may turn into a switch statement with cases for each React component. For now I'm trying to keep the changes limited to what will benefit the telephone migration.

bkjohnson committed:
Leave rule off by default for now

bkjohnson commented on .eslintrc.js:
```diff
@@ -88,6 +88,7 @@ module.exports = {
       },
     ],
     'va/correct-apostrophe': 1,
+    'va/prefer-web-component-library': 0,
```
This rule is off globally because if left on, husky will run the autofix on any changed file before commit and an error will be thrown since `no-unused-vars` will find unused imports:

![image](https://user-images.githubusercontent.com/2008881/150893472-d2c8691f-d254-4ba8-9989-596dfb17cec8.png)

We want to be careful not to be annoying if someone is trying to make a completely unrelated change. There's some background on this (and a slack thread) in https://github.com/department-of-veterans-affairs/vets-website/pull/17442.

bkjohnson commented:
I'm going to explore implementing this as a [suggestion](https://eslint.org/docs/developer-guide/working-with-rules#providing-suggestions) instead of an autofix.

bkjohnson committed:
Switch from autofix to suggestion

bkjohnson committed:
Turn warning back on

bkjohnson committed:
Strip hyphens out of contact

The new telephone web components expects only numbers for the contact
prop

bkjohnson committed:
Add unit tests for the rule

bkjohnson committed:
Add another test for notClickable transformation

bkjohnson committed:
Use regular booleans instead of ternarys

The filter at the end will take care of the false values

bkjohnson committed:
Only trigger on component-library imports

bkjohnson commented on script/eslint-plugin-va/rules/prefer-web-component-library.js:
```diff
@@ -0,0 +1,98 @@
+// Allows us to check if the JSX element we've identified is imported from
+// the component library. Some applications have their own components with the
+// same names as Design System components.
+// TODO: Consider improving the performance of this w/ memoization
+const isLibraryImport = (context, componentName) => {
+  let isLibraryComponent = false;
+  const allImports = context
+    .getAncestors()[0]
+    .body.filter(node => node.type === 'ImportDeclaration');
+  const componentLibraryImports = allImports.filter(node =>
+    node.source.value.includes(
+      '@department-of-veterans-affairs/component-library',
+    ),
+  );
+
+  componentLibraryImports.forEach(imp => {
+    if (imp.specifiers.find(i => i.local.name === componentName))
+      isLibraryComponent = true;
+  });
+
+  return isLibraryComponent;
+};
```
The performance on this probably isn't great considering we are currently running it on each JSX element in the syntax tree. The main goal here is to build migration functionality for `<Telephone>` - I think we can make improvements to the rule if we continue to develop it.

This is also less applicable for `<Telephone>` specifically, but I was doing this with `<AlertBox>` in mind, and there are a few others that fall into that category as well.

k80bowman replied:
yes, if there are ways to improve the performance of this, that would be nice. If it is not applicable for `<Telephone>` don't run it for now and maybe with the next iteration we can improve it.

bkjohnson committed:
Merge branch 'master' into migration-linter

bkjohnson committed:
Increase the version of our plugin

k80bowman commented on yarn.lock:
```diff
@@ -7986,8 +7986,9 @@ eslint-plugin-unicorn@^35.0.0:
-  version "1.0.5"
+"eslint-plugin-va@link:./script/eslint-plugin-va":
+  version "0.0.0"
+  uid ""
```
why is this going to version 0.0.0?

bkjohnson replied:
That happened when I switched the dependency protocol from `file` to `link`.

bkjohnson committed:
Remove check for component import source

bkjohnson committed:
Merge branch 'master' into migration-linter

bkjohnson committed:
Update rule to work on string literals

There's a new ESLint rule that's been turned on which is prompting teams
to not wrap string literals in curly braces. This changes the AST, so
now we've accounted for it.

bkjohnson committed:
Don't prefer single or double quotes in output

bkjohnson commented on script/eslint-plugin-va/rules/prefer-web-component-library.js:
```diff
@@ -9,7 +9,7 @@ const telephoneTransformer = (context, node) => {
   const notClickableNode = getPropNode(node, 'notClickable');
   const contactNode = getPropNode(node, 'contact');
-  const contactValue = contactNode?.value.expression;
+  const contactValue = contactNode?.value.expression || contactNode?.value;
```
The AST for the following has an `expression` node due to the curly braces:

```jsx
<Telephone contact={'12345'} />
```

And the AST for the following doesn't have that.

```jsx
<Telephone contact='12345' />
```

We are seeing more of that second pattern due to a recent linting rule that was turned on.

bkjohnson commented on script/eslint-plugin-va/rules/prefer-web-component-library.js:
```diff
@@ -31,9 +31,10 @@ const telephoneTransformer = (context, node) => {
-              fixer.replaceText(
-                contactValue,
-                `'${contactValue.value.replace(/[^\d]/g, '')}'`,
+              fixer.replaceTextRange(
+                // Leave the quotes alone
+                [contactValue.range[0] + 1, contactValue.range[1] - 1],
+                contactValue.value.replace(/[^\d]/g, ''),
```
This just replaces what's _inside_ the quotes rather than the entire quoted string. That way we don't have to worry about which quotes to use since certain rules can be picky about that.

bkjohnson reviewed (commented):
A few last-minute updates complete with test coverage.
